File: break-basic_SPN.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

from basic_SPN import *
import linear_cryptanalysis_lib as lc_lib
from math import fabs, ceil

# modify accordingly
def do_sbox(number):
    return sbox[number]

# modify accordingly
def do_inv_sbox(number):
    return sbox_inv[number]

# modify accordingly
def do_pbox(state):
    state_temp = 0
    for bitIdx in range(0,16):
        if(state & (1 << bitIdx)):
            state_temp |= (1 << pbox[bitIdx])
    state = state_temp
    return state

def main():

    NUM_P_C_PAIRS = 5000
    SBOX_BITS  = 4
    NUM_SBOXES = 4
    NUM_ROUNDS = 4
    MIN_BIAS = 0.008
    MAX_BLOCKS_TO_BF = 3

    lc_lib.initialize(NUM_P_C_PAIRS,
                      SBOX_BITS,
                      NUM_SBOXES,
                      NUM_ROUNDS,
                      MIN_BIAS,
                      MAX_BLOCKS_TO_BF,
                      do_sbox,
                      do_inv_sbox,
                      do_pbox)


    print('analizing cipher...')
    # there is no need to do this each time
    linear_aproximations = lc_lib.analize_cipher()

    print('\nbest 10 linear aproximations:')
    # just for demonstration
    for i in range(10):
        print(linear_aproximations[i])

    print('\nthe linear aproximation with the best bias will be used:')
    # you may choose anyone you like
    linear_aproximation = linear_aproximations[0]

    # just for demonstration
    end_sboxs = ', '.join(list(map(str, linear_aproximation[2])))
    print('ε: {:f}\nstart: sbox n°{:d}\nend sboxes:{}'.format(linear_aproximation[0], linear_aproximation[1][0], end_sboxs))
    Nl = ceil( pow(pow(linear_aproximation[0], -1), 2))
    print('\nneeded plaintext/ciphertext pairs (according to Matsui\'s paper):\nNl ≈ 1/(ε^2): {:d}'.format(ceil(Nl)))

    # this will be different with another cipher
    key = keyGeneration()
    # k is the last round key
    k = key[-3:]
    k = int(k, 16)

    # the 'encrypt' function might be different for you
    p_c_pairs = []
    for pt in range(NUM_P_C_PAIRS):
        p_c_pairs.append( [pt, encrypt(pt, key)] )

    print('\nbreaking cipher...\n')
    # obtain the biases given the p/c pairs and the linear aproximation
    biases = lc_lib.get_biases(p_c_pairs, linear_aproximation)

    # get the key with the highest bias
    maxResult, maxIdx = 0, 0
    for rIdx, result in enumerate(biases):
        if result > maxResult:
            maxResult = result
            maxIdx    = rIdx

    # maxIdx won't be equal to k if the final sboxes aren't consecutive!
    # in this example, they are
    if maxIdx == k:
        print('Success!')
        print('obtained key bits: {:d}'.format(maxIdx))
    else:
        print('Failure')


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        pass

File: C/break-basic_SPN.c
#include <stdio.h>
#include <stdint.h>
#include <sys/time.h>
#include <stdlib.h>
#include "basic_SPN.h"
#include "linear_cryptanalysis_lib.h"

// modify accordingly
uint16_t do_sbox(uint16_t number)
{
    return sbox[number];
}

// modify accordingly
uint16_t do_inv_sbox(uint16_t number)
{
    return sbox_inv[number];
}

// modify accordingly
uint16_t do_pbox(uint16_t state)
{
    uint16_t newSstate = 0;
    for (uint8_t bitIdx = 0; bitIdx < 16; bitIdx++)
    {
        if (state & (1 << bitIdx))
        {
            newSstate |= (1 << pbox[bitIdx]);
        }
    }
    return newSstate;
}

int main()
{
    struct timeval start, end, seed;
    printf("analizing cipher...\n\n");
    gettimeofday(&start, NULL);
    struct state** linear_aproximations = analize_cipher();
    gettimeofday(&end, NULL);
    printf("done. took %lu μs\n", (end.tv_sec - start.tv_sec) * 1000000 + end.tv_usec - start.tv_usec);

    if (linear_aproximations[0] == 0)
    {
        printf("no linear aproximations could be found!\n");
        exit(1);
    }

    printf("\nthe linear aproximation with the best bias will be used\n");
    struct state linear_aproximation = *linear_aproximations[0];
    printState(linear_aproximation);

    // key generation
    gettimeofday(&seed, NULL);
    srand(seed.tv_usec);
    uint8_t key[] = {rand(), rand(), rand(), rand(), rand(), rand(), rand(), rand(), rand(), rand()};

    // calculate the keybits that should be recovered
    uint16_t lastRoundKey = (key[8] << 8) | key[9];
    uint16_t keySearch = 0;
    uint8_t numKeyBlocks = 0;
    for (uint8_t sbox = 0; sbox < NUM_SBOXES; sbox++)
    {
        uint64_t input = bits_to_num(linear_aproximation.position[sbox]);
        if (input == 0) continue;
        numKeyBlocks++;
        uint16_t keyblock;
        keyblock = lastRoundKey >> ((NUM_SBOXES - sbox - 1) * SBOX_BITS);
        keyblock = keyblock & ((1 << SBOX_BITS) - 1);
        keyblock = keyblock << (((NUM_SBOXES - sbox - 1) * SBOX_BITS));
        keySearch |= keyblock;
    }

    // generate plaintext/ciphertext pairs
    uint64_t ciphertexts[NUM_P_C_PAIRS];
    uint64_t plaintexts[NUM_P_C_PAIRS];
    for (uint32_t i = 0; i < NUM_P_C_PAIRS; i++)
    {
        ciphertexts[i] = encrypt(i, key);
        plaintexts[i]  = i;
    }

    printf("\nbreaking cipher...\n\n");
    // obtain the biases given the p/c pairs and the linear aproximation
    gettimeofday(&start, NULL);
    double* biases = get_biases(plaintexts, ciphertexts, linear_aproximation);
    gettimeofday(&end, NULL);
    printf("done. took %lu μs\n", (end.tv_sec - start.tv_sec) * 1000000 + end.tv_usec - start.tv_usec);

    // get the key with the most hits
    double maxResult = 0;
    uint64_t maxIdx = 0;

    uint64_t maxKey = 1 << (numKeyBlocks * SBOX_BITS);
    for (int i = 0; i < maxKey; i ++)
    {
        if (biases[i] > maxResult)
        {
            maxResult = biases[i];
            maxIdx = i;
        }
    }

    if (maxIdx == keySearch)
    {
        printf("\nkey found: %d\n", maxIdx);
    }
    else
    {
        printf("\nFail!\n");
    }

    freeMem(linear_aproximations);
    free(biases);
    return 0;
}
